### PR TITLE
perf: selective activation checkpointing feature support 

### PR DESCRIPTION
**Pull request by seonjinn**
Supports Megatron-Core recompute_granularity and recompute_modules through policy config and megatron setup so training can selectively recompute activations (core_attn, moe, moe_act, etc.) instead of full checkpointing.

# What does this PR do ?

**Add a one line overview of what this PR aims to accomplish.**

# Issues
List issues that this PR closes ([syntax](https://docs.github.com/en/issues/tracking-your-work-with-issues/using-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword)):

# Usage
* **You can potentially add a usage example below**

```python
# Add a code snippet demonstrating how to use this
```

# Before your PR is "Ready for review"
**Pre checks**:
- [ ] Make sure you read and followed [Contributor guidelines](/NVIDIA-NeMo/RL/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you run the unit tests and functional tests locally? Visit our [Testing Guide](/NVIDIA-NeMo/RL/blob/main/docs/testing.md) for how to run tests
- [ ] Did you add or update any necessary documentation? Visit our [Document Development Guide](/NVIDIA-NeMo/RL/blob/main/docs/documentation.md) for how to write, build and test the docs.

# Additional Information
* ...

### CODE REVIEW

File: nemo_rl/models/megatron/setup.py [hunk 1]
```diff
@@ -468,9 +468,21 @@ def _apply_performance_config(model_cfg: Any, config: PolicyConfig) -> None:
 
     # Activation checkpointing
     if config["megatron_cfg"]["activation_checkpointing"]:
-        model_cfg.recompute_granularity = "full"
-        model_cfg.recompute_method = "uniform"
-        model_cfg.recompute_num_layers = 1
+        granularity = config["megatron_cfg"].get("recompute_granularity", "full")
```
Comment (terrykong) [inline nemo_rl/models/megatron/setup.py]: [`setup.py:471`](https://github.com/NVIDIA-NeMo/RL/blob/6621d76218009f065e4622e03a3f34b3398768fe/nemo_rl/models/megatron/setup.py#L471)

The `.get("recompute_granularity", "full")` pattern uses a hidden non-`None` default. The repo's [config-conventions skill](https://github.com/NVIDIA-NeMo/RL/blob/6621d76218009f065e4622e03a3f34b3398768fe/.claude/skills/config-conventions/SKILL.md) (§"Accessing NotRequired Fields") says: *"Never provide a non-`None` default."* [CONTRIBUTING.md](https://github.com/NVIDIA-NeMo/RL/blob/6621d76218009f065e4622e03a3f34b3398768fe/CONTRIBUTING.md) reiterates: *"Do not add arbitrary defaults for configs, be as explicit as possible."*

Since the exemplar YAML already ships `recompute_granularity: "full"`, the key should always be present — access it directly like sibling fields (e.g. `activation_checkpointing`):

```suggestion
        granularity = config["megatron_cfg"]["recompute_granularity"]
```

File: nemo_rl/models/megatron/setup.py [hunk 2]
```diff
@@ -468,9 +468,21 @@ def _apply_performance_config(model_cfg: Any, config: PolicyConfig) -> None:
...
+        model_cfg.recompute_granularity = granularity
+        if granularity == "full":
+            model_cfg.recompute_method = "uniform"
+            model_cfg.recompute_num_layers = 1
+        elif granularity == "selective":
+            recompute_modules = config["megatron_cfg"].get("recompute_modules")
+            if recompute_modules is not None:
```
Comment (terrykong) [inline nemo_rl/models/megatron/setup.py:479]: [`setup.py:478`](https://github.com/NVIDIA-NeMo/RL/blob/6621d76218009f065e4622e03a3f34b3398768fe/nemo_rl/models/megatron/setup.py#L478)

There’s a validation gap here worth noting. Here’s the call chain:

1. [`MegatronPolicyWorker.__init__`](https://github.com/NVIDIA-NeMo/RL/blob/6621d76218009f065e4622e03a3f34b3398768fe/nemo_rl/models/policy/workers/megatron_policy_worker.py#L150) calls [`validate_and_set_config()`](https://github.com/NVIDIA-NeMo/RL/blob/6621d76218009f065e4622e03a3f34b3398768fe/nemo_rl/models/megatron/setup.py#L157).

2. That calls [`setup_model_config()`](https://github.com/NVIDIA-NeMo/RL/blob/6621d76218009f065e4622e03a3f34b3398768fe/nemo_rl/models/megatron/setup.py#L223), which loads the pretrained config via [`ConfigContainer.from_yaml()`](https://github.com/NVIDIA-NeMo/RL/blob/6621d76218009f065e4622e03a3f34b3398768fe/nemo_rl/models/megatron/setup.py#L308). This constructs the `TransformerConfig` dataclass, and [`__post_init__`](https://github.com/NVIDIA/Megatron-LM/blob/d30c3ae5469fe3f6a64d4fd2e63b6e7f7844ea81/megatron/core/transformer/transformer_config.py#L1363-L1378) runs — validating `recompute_modules` against the [allowed set](https://github.com/NVIDIA/Megatron-LM/blob/d30c3ae5469fe3f6a64d4fd2e63b6e7f7844ea81/megatron/core/transformer/transformer_config.py#L1365-L1373).

3. The validated config is stored as [`model_cfg = cfg_from_pretrained.model`](https://github.com/NVIDIA-NeMo/RL/blob/6621d76218009f065e4622e03a3f34b3398768fe/nemo_rl/models/megatron/setup.py#L325).

4. Later, [`_apply_performance_config(model_cfg, config)`](https://github.com/NVIDIA-NeMo/RL/blob/6621d76218009f065e4622e03a3f34b3398768fe/nemo_rl/models/megatron/setup.py#L341) sets [`model_cfg.recompute_modules`](https://github.com/NVIDIA-NeMo/RL/blob/6621d76218009f065e4622e03a3f34b3398768fe/nemo_rl/models/megatron/setup.py#L479) via plain attribute assignment. This does **not** re-trigger `__post_init__`, so the validation from step 2 is bypassed.

A typo like `["mopee"]` would be silently accepted. At runtime, MCore checks module names via string matching (e.g. [`"core_attn" in self.config.recompute_modules`](https://github.com/NVIDIA/Megatron-LM/blob/d30c3ae5469fe3f6a64d4fd2e63b6e7f7844ea81/megatron/core/transformer/transformer_block.py)), so an unrecognized name is a silent no-op — no memory is saved despite the user’s intent.

Could you add a comment here noting that validation of `recompute_modules` values happens in [MCore’s `TransformerConfig.__post_init__`](https://github.com/NVIDIA/Megatron-LM/blob/d30c3ae5469fe3f6a64d4fd2e63b6e7f7844ea81/megatron/core/transformer/transformer_config.py#L1365-L1373) and that valid values are: `core_attn`, `moe_act`, `layernorm`, `mla_up_proj`, `mlp`, `moe`, `shared_experts`? Filed [#2291](https://github.com/NVIDIA-NeMo/RL/issues/2291) to track refactoring the config init flow so we can leverage `__post_init__` validation after these mutations.

```suggestion
            recompute_modules = config["megatron_cfg"].get("recompute_modules")
            if recompute_modules is not None:
                # NOTE: MCore validates recompute_modules in TransformerConfig.__post_init__,
                # but that validation doesn't re-run after attribute assignment here.
                # Valid values: core_attn, moe_act, layernorm, mla_up_proj, mlp, moe, shared_experts
                # See: https://github.com/NVIDIA/Megatron-LM/blob/d30c3ae5469fe3f6a64d4fd2e63b6e7f7844ea81/megatron/core/transformer/transformer_config.py#L1365
                # Tracking: https://github.com/NVIDIA-NeMo/RL/issues/2291
                model_cfg.recompute_modules = recompute_modules
```

File: nemo_rl/models/policy/__init__.py [hunk 3]
```diff
@@ -203,6 +203,14 @@ class MegatronConfig(TypedDict):
     # Setting to 0 is faster, but you are more likely to run out of GPU memory. In SFT/DPO, the default is 0.
     empty_unused_memory_level: int
     activation_checkpointing: bool
+    # Recompute granularity: "full" recomputes all activations, "selective" recomputes
+    # only specific modules (see recompute_modules). "selective" typically saves ~10-18GB
+    # for MoE models while retaining higher throughput than "full".
+    recompute_granularity: NotRequired[Literal["full", "selective"]]
+    # Modules to selectively recompute when recompute_granularity="selective".
+    # MCore valid options: ["core_attn", "moe", "mla_up_proj"]. Defaults to ["core_attn"]
```
Comment (terrykong) [inline nemo_rl/models/policy/__init__.py]: [`__init__.py:211`](https://github.com/NVIDIA-NeMo/RL/blob/6621d76218009f065e4622e03a3f34b3398768fe/nemo_rl/models/policy/__init__.py#L211)

The comment lists 3 valid modules but MCore allows 7. From [`TransformerConfig.recompute_modules`](https://github.com/NVIDIA/Megatron-LM/blob/d30c3ae5469fe3f6a64d4fd2e63b6e7f7844ea81/megatron/core/transformer/transformer_config.py#L483-L486):

> choices: "core_attn", "moe_act", "layernorm", "mla_up_proj", "mlp", "moe", "shared_experts"

[Validation at L1365–1373](https://github.com/NVIDIA/Megatron-LM/blob/d30c3ae5469fe3f6a64d4fd2e63b6e7f7844ea81/megatron/core/transformer/transformer_config.py#L1365-L1373) enforces this set. Could you update to list all 7 or link to MCore?

```suggestion
    # MCore valid options: ["core_attn", "moe_act", "layernorm", "mla_up_proj", "mlp", "moe", "shared_experts"].
    # Defaults to ["core_attn"] when None. Full list and per-module constraints:
    # https://github.com/NVIDIA/Megatron-LM/blob/d30c3ae5469fe3f6a64d4fd2e63b6e7f7844ea81/megatron/core/transformer/transformer_config.py#L483
```

File: examples/configs/grpo_math_1B.yaml [hunk 4]
```diff
@@ -131,6 +131,15 @@ policy:
     force_reconvert_from_hf: False # Set to True to force reconvert of the model from Hugging Face
     empty_unused_memory_level: 1  # 1 is the minimum recommendation for RL since we almost always need to offload before beginning generation. Setting to 0 is faster, but you are more likely to run out of GPU memory.
     activation_checkpointing: false
+    # recompute_granularity controls activation checkpointing depth.
+    # "full": recompute all activations (default, max memory savings).
+    # "selective": recompute only specific modules (see recompute_modules).
+    # Only takes effect when activation_checkpointing: true.
+    recompute_granularity: "full"
+    # Modules to selectively recompute when recompute_granularity="selective".
+    # Options: ["core_attn", "moe", "mla_up_proj"]. Null defaults to ["core_attn"].
```
Comment (terrykong) [inline examples/configs/grpo_math_1B.yaml]: [`grpo_math_1B.yaml:140`](https://github.com/NVIDIA-NeMo/RL/blob/6621d76218009f065e4622e03a3f34b3398768fe/examples/configs/grpo_math_1B.yaml#L140)

Same as the TypedDict comment — lists 3 modules but MCore allows 7. Consider updating:

```suggestion
    # MCore options: ["core_attn", "moe_act", "layernorm", "mla_up_proj", "mlp", "moe", "shared_experts"].
    # Null defaults to ["core_attn"]. Full list and per-module constraints:
    # https://github.com/NVIDIA/Megatron-LM/blob/d30c3ae5469fe3f6a64d4fd2e63b6e7f7844ea81/megatron/core/transformer/transformer_config.py#L483
```

**Review by terrykong — COMMENTED** [review]
## Review Summary

Thanks for adding selective activation checkpointing support — this is a valuable feature that exposes a real MCore knob MoE users need. The implementation is well-scoped and the branching logic is clean.

### Suggestions

**Config-conventions guideline** (inline): The `.get("recompute_granularity", "full")` pattern uses a hidden non-`None` default, which the repo guidelines forbid. The key should be accessed directly.

**Incomplete module list** (inline): The `recompute_modules` comments list 3 valid options but MCore allows 7.

**Validation gap** (inline): Invalid `recompute_modules` values (typos) are silently accepted because MCore's `__post_init__` validation doesn’t re-run after attribute assignment. Filed #2291 to track a broader refactor.

### Performance Evidence

Since this is a `perf:`-labeled PR and the TypedDict comment claims "∼10–18GB savings for MoE models," could you share a before/after comparison? For example, peak GPU memory and tokens/sec with `selective` vs `full` recompute on a representative MoE model would help future users understand the expected benefit and source the claim. Filling in the PR template sections would also be helpful.

### Test Coverage (nit)

The new `selective` branch and `ValueError` path have no unit test coverage. Consider adding tests for: (a) selective with custom modules, (b) selective with `None` modules (MCore default), (c) invalid granularity raises `ValueError`. Happy to share three pre-verified tests that match the existing MagicMock-based style.

### Linter: PASS
All hooks passed (ruff, ruff-format, taplo, pyrefly, end-of-files, trailing-whitespace, minimize-check).

_Generated by Claude Code_

**Comment by seonjinn** [comment]
/ok to [150ab01](https://github.com/NVIDIA-NeMo/RL/pull/2280/commits/150ab0143e8b18142c2f75a3db8840b0c1ca63c9)

File: nemo_rl/models/megatron/setup.py (hunk 2 above)
Comment (seonjinn) [reply nemo_rl/models/megatron/setup.py:479]: Thank you for suggestion!

**Comment by seonjinn** [comment]
/ok to test [c7db964](https://github.com/NVIDIA-NeMo/RL/pull/2280/commits/c7db96489750cf18b52cba890c219096843f11dc)

**Comment by seonjinn** [comment]
@terrykong  I've made changes based on your feedback. Does this PR look good to you ?

**Comment by seonjinn** [comment]
/ok to test [0136f76](https://github.com/NVIDIA-NeMo/RL/pull/2280/commits/0136f768eaafd25bacb26c55db320a55e30662b6)

**Comment by seonjinn** [comment]
/ok to test [bd4b112](https://github.com/NVIDIA-NeMo/RL/pull/2280/commits/bd4b1122898ce51a9ac897d2c42ec18e3535f7fb)

**Comment by seonjinn** [comment]
@terrykong can you review this PR ?

File: examples/configs/grpo_math_1B.yaml [hunk 5]
```diff
@@ -134,6 +134,17 @@ policy:
     force_reconvert_from_hf: False # Set to True to force reconvert of the model from Hugging Face
     empty_unused_memory_level: 1  # 1 is the minimum recommendation for RL since we almost always need to offload before beginning generation. Setting to 0 is faster, but you are more likely to run out of GPU memory.
     activation_checkpointing: false
+    # recompute_granularity controls activation checkpointing depth.
+    # "full": recompute all activations (default, max memory savings).
+    # "selective": recompute only specific modules (see recompute_modules).
+    # Only takes effect when activation_checkpointing: true.
+    recompute_granularity: "full"
```
[the remark this answers is not in the document]
Comment (terrykong) [reply examples/configs/grpo_math_1B.yaml:141]: @seonjinn i do agree with claude here that this should probably be repeated in the other exemplar configs

**Review by terrykong — CHANGES_REQUESTED** [review]
lgtm. just that one comment about the other exemplar configs

**Review by terrykong — APPROVED** [review]
